data_structures/src/data_request.rs

```diff
@@ -412,69 +425,87 @@ where
     if let Stage::Tally(tally_metadata) = &report.metadata {
         let commits_count = committers.len();
         let reveals_count = revealers.len();
-        let mut slashed_witnesses = committers;
+        let mut out_of_consensus = committers;
+        let mut error_committers = vec![];
 
         let liars = &tally_metadata.liars;
+        let errors = &tally_metadata.errors;
 
-        if reveals_count != liars.len() {
-            return Err(TransactionError::MismatchingLiarsNumber {
-                reveals_n: reveals_count,
-                inputs_n: liars.len(),
-            }
-            .into());
-        }
+        assert_eq!(reveals_count, liars.len(), "Length of liars vector collected from tally ({}) does not match actual count of reveals ({})", reveals_count, liars.len());
+        assert_eq!(reveals_count, errors.len(), "Length of errors vector collected from tally ({}) does not match actual count of reveals ({})", reveals_count, errors.len());
+
+        let (liars_count, errors_count) = calculate_errors_and_liars_count(errors, liars);
+
+        let collateral = if dr_output.collateral == 0 {
+            collateral_minimum
+        } else {
+            dr_output.collateral
+        };
 
-        let liars_count = liars.iter().fold(0, |count, liar| match liar {
-            true => count + 1,
-            false => count,
-        });
-        let honests_count = reveals_count - liars_count;
         // Collateral division rest goes for the miner
         let (reward, _rest) = calculate_witness_reward(
             commits_count,
             reveals_count,
-            honests_count,
-            &dr_output,
-            collateral_minimum,
+            liars_count,
+            errors_count,
+            dr_output.witness_reward,
+            collateral,
         );
 
         let mut outputs: Vec<ValueTransferOutput> = if reveals_count > 0 {
             revealers
                 .iter()
-                .zip(liars.iter())
-                .filter_map(|(&revealer, &liar)| {
-                    // Only reward reveals in consensus
-                    if liar {
-                        None
-                    } else {
+                .zip(liars.iter().zip(errors.iter()))
+                .filter_map(|(&revealer, x)| match x {
```

Comment:
`match (liar, error)` would be easier to understand, now it's confusing, what's `(true, false)`?

And does it make sense to order the branches somehow? For example `(false, false)` first, `(true, true)` last, or vice versa.